pytket/extensions/cutensornet/mps/mps_gate.py

```diff
@@ -0,0 +1,336 @@
+
+        # Create tensor descriptors
+        T_desc = T.get_tensor_descriptor(self._libhandle)
+        L_desc = L.get_tensor_descriptor(self._libhandle)
+        R_desc = R.get_tensor_descriptor(self._libhandle)
+
+        # Create SVDConfig with default configuration
+        svd_config = cutn.create_tensor_svd_config(self._libhandle)
+        # Create SVDInfo to record truncation information
+        svd_info = cutn.create_tensor_svd_info(self._libhandle)
+
+        if self.truncation_fidelity < 1:
+            # Carry out SVD decomposition first with NO truncation
+            # to figure out where to apply the dimension cutoff.
+            #
+            # NOTE: Unfortunately, it looks like there is no option
+            # in the cuTensorNet API to ask for a cutoff in terms of
+            # `discarded_weight`; ABS_CUTOFF and REL_CUTOFF aren't it.
+            # It'd be best to request them to add this feature.
+            #
+            # TODO: Benchmark if this workflow that does two SVDs on
+            # the same tensor is slower than if I were to do the
+            # tensor reshaping, normalisation of S and contraction myself.
+            # It's likely, since SVD is costly, but this is simpler.
+
+            cutn.tensor_svd(
+                self._libhandle,
+                T_desc,
+                T.data.data.ptr,
+                L_desc,
+                L.data.data.ptr,
+                S_d.data.ptr,
+                R_desc,
+                R.data.data.ptr,
+                svd_config,
+                svd_info,
+                0,  # 0 means let cuQuantum manage mem itself
+                self._stream.ptr,  # type: ignore
+            )
+            self._stream.synchronize()  # type: ignore
+
+            # Use the fact that the entries of S_d are sorted in decreasing
+            # order and calculate the number of singular values `new_dim` to
+            # keep so that
+            #                             sum([s**2 for s in S'])
+            #   truncation_fidelity  <=  -------------------------
+            #                             sum([s**2 for s in S])
+            #
+            # where S is the list of original singular values and S' is the set of
+            # singular values that remain after truncation (before normalisation).
+            denom = float(sum(cp.square(S_d)))  # Element-wise squaring
+            numer = 0.0
+            new_dim = 0
+
+            while self.truncation_fidelity > numer / denom:
+                numer += float(S_d[new_dim] ** 2)
+                new_dim += 1
+
+            # We need to change the shape of the tensors
+            l_shape[-2] = new_dim
+            L.data = cp.empty(l_shape, dtype=self._complex_t)
+            r_shape[0] = new_dim
+            R.data = cp.empty(r_shape, dtype=self._complex_t)
+            S_d = cp.empty(new_dim, dtype=self._real_t)
+
+            # Create new tensor descriptors
+            cutn.destroy_tensor_descriptor(L_desc)

```

Comment:
My thinking was that, since any user working with tensor descriptors is very likely to also be using other handles such as those for `svd_info` and `svd_config` (or equivalent for QR), it's fine to ask them to destroy handles using `cuTensorNet`. 

I think adding these methods to `Tensor` is just abstraction for the sake of abstraction, which IMO is unnecessary. It'd mean keeping track of the handle within `Tensor`, but the handle outside this functions is always `None`.

Reply:
In principle keeping track of handles (pointers) automatically is a good thing and prevents memory leaks, but as far as I understand here we can't really do something like that (like a smart pointer in C++), can we? In that case perhaps it is indeed fine as it is now.

Reply:
Yeah, but if we just wrap `destroy_tensor_descriptor` in our `Tensor` class we are not solving that problem. Best would be only create/destroy tensors inside methods encapsulating things like SVD but, as discussed in my comment below, although I think that's reasonable, I think we shouldn't do it for now.

Reply:
BTW, I think using the `Network` object takes care of a good amount of this tensor descriptor nonsense. That might be the best way to go, but that'll come in a different PR.

Reply:
Agreed!